### ISSUE TITLE
[Feature request] To be able to intercept module calls for an existing model.

### ISSUE THREAD

**Issue by zpolina** [google/flax#1355]
We are interested in enabling a ‘hook’(‘interceptor’) into calls to Flax modules to add functionality for various purposes. For example, hooks would make it possible to intercept module function calls by changing the input or perturbing the output. It can also be used to access or modify the module gradients. The main motivation for this is to avoid forking existing off-the-shelf  models and create code duplication when it could be avoided. It is also not always feasible to make changes to existing models. For example, a user might want to access a certain intermediate output or gradient for a large number of models. 

Main use cases for this functionality are about perturbing/pruning models or explaining models behavior by looking into gradients. Often the perturbation parameters can be learnable, and injected into model particular layers. For example:

- Differential Masking ([ArXiv](https://arxiv.org/abs/2004.14992)) ([GitHub example with a hook](https://github.com/nicola-decao/diffmask/blob/2385532d4859d6b1608892afe0f827c726aeda75/diffmask/utils/getter_setter.py#L76))
- Parameter space noise for exploration ([ArXiv](https://arxiv.org/pdf/1706.01905.pdf))  ([GitHub example with a hook](https://github.com/iffiX/machin/blob/75b271fb0384a986a6012e17764ceaad65550915/machin/frame/noise/param_space_noise.py#L110))
- SparseML ([Documentation](https://docs.neuralmagic.com/sparseml/))  ([GitHub example with a hook](https://github.com/neuralmagic/sparseml/blob/7c72143a6a8501c7271aa66695be9c0a35c3abb5/src/sparseml/pytorch/optim/mask_pruning.py#L652))
- Class Activation Mapping ([ArXiv](https://arxiv.org/abs/1512.04150))  ([GitHub example with a hook](https://github.com/Project-MONAI/MONAI/blob/12f267c98eabdcd566dff11a3daf931201f04da4/monai/visualize/class_activation_maps.py#L77))

  Note that the hook is a generalization of `capture_intermediates`, in the sense that `capture_intermediates` can be implemented as an interceptor as in the following example for the case of `capture_intermediates=True`:

```
def intercept_method(mdl, fun, *args, **kwargs):
  if fun.__name__ == '__call__':
    y = fun(mdl, *args, **kwargs)
    mdl.sow('intermediates', fun.__name__, y)
    return y
  return fun(mdl, *args, **kwargs)
```
It might be convenient for users if `capture_intermediates` stays, but it could be implemented as an intercept method.

Other frameworks:

- [PyTorch allows](https://pytorch.org/docs/stable/generated/torch.nn.Module.html?highlight=register_forward_hook#torch.nn.Module.register_forward_hook) a user to register hooks using `register_forward_hook`  and `register_forward_pre_hook`. It also has a `register_full_backward_hook` that is called during the backward pass. 
- In the JAX ecosystem, Haiku has an `intercept_methods` [function](https://dm-haiku.readthedocs.io/en/latest/api.html?highlight=getter#haiku.experimental.intercept_methods).

**Comment by jheek** [comment]
Thanks for your proposal! Hooks are indeed very powerful and seem to be used quite a bit in the PyTorch ecosystem.
Historically, we have been hesitant to add hooks because they allow for injecting arbitrary side-effects at function boundaries unlike the "passive observing" done by `capture_intermediates`.
We decided that a big feature like this needs a [FLIP](https://github.com/google/flax/tree/master/docs/flip) to carefully weigh the APIs decisions, pros, and cons.

So far we have had some reservations about adding hooks and we re-discussed the idea in our last team meeting. This is a summary of points that were raised:
- What happens if hooks and transforms (e.g.: `nn.scan`, `nn.vmap`) act on the same method?
- How to compose multiple hooks?
- Should it be possible to add hooks at multiple levels (so not only at nn.apply)?
- Why does linen need a custom hook solution? There are various libraries and patterns in Python that allow hooks ranging from simple duck typing to fancy Aspect Oriented Programming (AOT) libraries (http://pytilities.sourceforge.net/doc/1.2.0/guide/aop/).

**Comment by marcvanzee** [comment]
Hi @zpolina @bastings, any updates on this? I think converting this to a FLIP would be nice so we can discuss it in more detail.

To give an example of the usefulness of FLIPS: we had some vague ideas of using Optax for our optimizers, but there were a number of unanswered questions. @andsteing started FLIP #1009 , which lead to a detailed discussion that helped us to resolve all the questions, and we are now using Optax in our examples.

**Comment by zpolina** [comment]
Hi @marcvanzee, thank you for checking on. We are about to start writing it and will get in touch soon.

This is a great positive example! Thank you for sharing. Hopefully it will also help us answer all the questions.

**Comment by marcvanzee** [comment]
@zpolina any updates on this?

**Comment by marcvanzee** [comment]
Update: I just noticed that there is a FLIP for this, so I am closing this issue in favor of the FLIP: #1436
